src/parser/SparqlParser.cpp

```diff
@@ -9,379 +9,472 @@
                                " is selected but not "
-                               "aggregated despite the result not being"
+                               "aggregated despite the query not being "
                                "grouped by " +
-                               var + ".");
+                               var + ".\n" + _lexer.input());
         }
       }
     }
   }
-  result.parseAliases();
-  return result;
+
+  ad_utility::HashMap<std::string, size_t> variable_counts;
+  for (const std::string& s : query->_selectedVariables) {
+    variable_counts[s]++;
+  }
+  for (const ParsedQuery::Alias& a : query->_aliases) {
+    // The variable was already added to the selected variables while
+    // parsing the alias, thus it should appear exactly once
+    if (variable_counts[a._outVarName] > 1) {
+      throw ParseException("The variable name " + a._outVarName +
+                           " used in "
+                           "an alias was already selected on.\n" +
+                           _lexer.input());
+    }
+    variable_counts[a._outVarName];
+  }
 }
 
 // _____________________________________________________________________________
-void SparqlParser::parsePrologue(string str, ParsedQuery& query) {
-  if (str.find("BASE") != string::npos) {
-    throw ParseException("Bases are not supported, yet.");
+void SparqlParser::parsePrologue(ParsedQuery* query) {
+  while (_lexer.accept("prefix")) {
+    _lexer.expect(SparqlToken::Type::IRI);
+    string key = _lexer.current().raw;
+    _lexer.expect(SparqlToken::Type::IRI);
+    string value = _lexer.current().raw;
+    addPrefix(key, value, query);
   }
+}
 
-  size_t i = str.find("PREFIX");
-  while (i != string::npos) {
-    size_t j = str.find("PREFIX", i + 1);
-    addPrefix(str.substr(i, j - i), query);
-    i = j;
-  }
+// _____________________________________________________________________________
+void SparqlParser::addPrefix(const string& key, const string& value,
+                             ParsedQuery* query) {
+  // Remove the trailing : from the key
+  SparqlPrefix p{key.substr(0, key.size() - 1), value};
+  query->_prefixes.emplace_back(p);
 }
 
 // _____________________________________________________________________________
-void SparqlParser::addPrefix(const string& str, ParsedQuery& query) {
-  auto parts = ad_utility::splitWs(str);
-  if (parts.size() != 3) {
-    throw ParseException(string("Invalid PREFIX statement: ") + str);
-  }
-  string uri = ad_utility::strip(parts[2], " \t\n");
-  if (uri.size() == 0 || uri[0] != '<' || uri[uri.size() - 1] != '>') {
-    throw ParseException(string("Invalid URI in PREFIX: ") + uri);
+void SparqlParser::parseSelect(ParsedQuery* query) {
+  if (_lexer.accept("distinct")) {
+    query->_distinct = true;
+  }
+  if (_lexer.accept("reduced")) {
+    query->_reduced = true;
+  }
+  while (!_lexer.accept("where")) {
+    if (_lexer.accept(SparqlToken::Type::VARIABLE)) {
```

Comment:
I think the problem with the last query is that you need to accept `"*"` here too.

Reply:
And it doesn't cause an exception because `"*"` is a SPARQL symbol and thus `readNext()` succeeds.

Reply:
@niklas88 I added validation to the parsing of the variables in the select clause

Reply:
I've got a fix for this prepared, will push to the PR branch as soon as I've tried with more queries.